### ISSUE TITLE
Windows Server 2019 KB4551853 Update breaks Docker Swarm Ingress network

### ISSUE THREAD

**Issue by djarvis** [moby/moby#40998]
Windows Server 2019 running Docker/Swarm, ingress network was working fine until this was installed:

2020-05 Cumulative Update for Windows Server 2019 (1809) for x64-based Systems (KB4551853)

This broke something with the ingress network such that no traffic could enter through any exposed/published ports, either from the local machine or remotely.

Uninstalling this update made it all work again.

Deploying containers on the nat network worked fine.  Access from the containers to the external network remained working in a docker swarm, but no traffic could come in.

**Comment by simmohall** [comment]
yep we just had same problem. Our services within the swarm couldn't communicate with each other either. :( :(

**Comment by veepee78** [comment]
I had issues already with 2020-04 cumulative update. 

Currently my swarm status is that containers using overlay network can contact services running in our network like mq and db, but if you go inside the container and try to use tracert or ping, they can't connect anything!? Also, if I expose port from container to host, it does not work. 

Ingress network is somehow messed up. Does not help to delete all hns networks and let docker create them again. Same situation continues.

Nat network works.

**Comment by alexftr** [comment]
Same problem for us. Containers can't communicate with each other. 
@djarvis How did you uninstall that package? 
I tried `wusa /uninstall /kb:4551853` 
but it it said  "...required by your computer and cannot be uninstalled"

**Comment by avinhan** [comment]
@alexftr we have the same issue with a couple of our environments in Azure. We had the issue happen to one of our dev environments a week back. But at that time we were able to uninstall it. Today it happened in our QA environments, but now we cannot uninstall it. We get the following error:

![image](https://user-images.githubusercontent.com/17634705/83009555-45fe0e80-a05a-11ea-9be9-88199b5e7220.png)

Did anyone manage to fix the Swarm network issue?

**Comment by djarvis** [comment]
@alexftr Same story as @avinhan, we were able to freely uninstall it a few weeks ago but now we are not.  Looking for a fix to this or a way to rip out this update that seems to now be cemented in.

**Comment by jabteles** [comment]
Having the same issue. Uninstalled and now I can access other services by name from other service.

**Comment by djarvis** [comment]
@jabteles How did you uninstall the update?

**Comment by EagleIJoe** [comment]
Uninstalled with the above wusa command did the trick on my workers.

FWIW: it is not _only_ swarm related. I had issues with my kubernetes / flannel / docker workers.

**Comment by jabteles** [comment]
> @jabteles How did you uninstall the update?

Start > write "update"  > click "Windows Update Settings" > click "View update history" > click "Uninstall updates" > filter "KB551853" in the search field > select update and click Uninstall

**Comment by thaJeztah** [comment]
ping @daschott @dperny @taylorb-microsoft ptal

**Comment by djarvis** [comment]
> > @jabteles How did you uninstall the update?
> 
> Start > write "update" > click "Windows Update Settings" > click "View update history" > click "Uninstall updates" > filter "KB551853" in the search field > select update and click Uninstall

Ok.  Yes.  I was able to do that on one machine a few weeks ago but now and on other machines it seems the update cannot be uninstalled so easily.

**Comment by jabteles** [comment]
> > > @jabteles How did you uninstall the update?
> > 
> > 
> > Start > write "update" > click "Windows Update Settings" > click "View update history" > click "Uninstall updates" > filter "KB551853" in the search field > select update and click Uninstall
> 
> Ok. Yes. I was able to do that on one machine a few weeks ago but now and on other machines it seems the update cannot be uninstalled so easily.

What is the error? Maybe a reboot is necessary before?

**Comment by djarvis** [comment]
> What is the error? Maybe a reboot is necessary before?

There is no "uninstall" link in the UI, and this command:

wusa /uninstall /kb:4551853

comes up with some "...required by your computer and cannot be uninstalled" message.

Apparently if you create a Windows Server 2019 VM in Azure today the KB is already installed and is unable to be removed.  On VMs created months ago I was able to uninstall this update and prevent it from being installed.

**Comment by jabteles** [comment]
> > What is the error? Maybe a reboot is necessary before?
> 
> There is no "uninstall" link in the UI, and this command:
> 
> wusa /uninstall /kb:4551853
> 
> comes up with some "...required by your computer and cannot be uninstalled" message.
> 
> Apparently if you create a Windows Server 2019 VM in Azure today the KB is already installed and is unable to be removed. On VMs created months ago I was able to uninstall this update and prevent it from being installed.

I understand, can't help on that, I'm using on premises VM's and it was possible.

**Comment by daschott** [comment]
Thanks for reporting this issue, I will raise this issue with the team for further investigation. If someone has a repro and could be kind enough to run https://github.com/microsoft/SDN/blob/master/Kubernetes/windows/debug/collectlogs.ps1 and share the service/container names, that would be helpful in accelerating resolution.

**Comment by alexftr** [comment]
@djarvis @avinhan Yes, i couldn't uninstall that update. I had to recreate recreate VMs in Azure from snapshot from a month ago. This was the only option for me.

**Comment by avinhan** [comment]
Hi @daschott any updates on a resolution / workaround?

**Comment by jabteles** [comment]
@avinhan workaround is to uninstall the update. Worked for me

**Comment by tfenster** [comment]
@daschott are you still looking for debug logs? I have a fully automated repro

**Comment by daschott** [comment]
@tfenster yes please. Unfortunately this is not reproducing for me and others that have tried on a Windows Server 2019 machine with latest hotfix. Perhaps can someone share a sample .yaml file as well in case there is something different in the configuration?

**Comment by tfenster** [comment]
@daschott I tried to narrow it down to a simpler scenario and there it doesn't break anymore. I'll need to expand until I get to my more complex scenario to find out where it breaks. If you want, you could clone https://github.com/tfenster/bc-swarm and then follow https://raw.githubusercontent.com/tfenster/BC-Swarm/master/BC/howto.md, which definitely breaks. But I'll try to simplify that, but probably will take me 24-48h to get there

**Comment by simmohall** [comment]
@daschott  I can reproduce this morning by spinning up 2 azure vms (--image Win2019Datacenter).
Do the needful to create the swarm (docker swarm init, then join..)
Then spin up some dummy stack using below assuming yml below in file test.yml
`docker stack deploy --compose-file test.yml test`
Then try ping 1 service from the other (or curl). both commands fail/timeout
```
docker exec <containerIdofWeb> ping api
docker exec <containerIdofWeb> curl http://api
```

```
version: "3.7"
services:
  web:
    image: mcr.microsoft.com/dotnet/core/samples:aspnetapp-nanoserver-1809
    ports:
      - "80:80"
    deploy:
      mode: replicated
      restart_policy:
        condition: on-failure
  api:
    image: mcr.microsoft.com/dotnet/core/samples:aspnetapp-nanoserver-1809
    ports:
      - "8080:80"
    deploy:
      mode: replicated
      restart_policy:
        condition: on-failure
```

**Comment by tfenster** [comment]
@daschott It was a bit more complex than I though, but here is the simplest scenario I could come up with where it breaks. It needs two networks and a reboot, but then it realiably breaks for me with newer Azure VM images while it still works with those from April 2020. Here are my repro steps:

1. go to shell.azure.com
2. switch to PowerShell
3. run the following script to create a manager and a worker vm, based on https://docs.microsoft.com/en-us/azure/virtual-machines/scripts/virtual-machines-windows-powershell-sample-create-vm
```
# Variables for common values
$resourceGroup = "repro-swarm-bug"
$location = "westeurope"
$mgrName = "mgr"
$workerName = "worker"

# Create user object
$cred = New-Object System.Management.Automation.PSCredential ("vmadmin", (ConvertTo-SecureString -String "Passw0rd*123" -AsPlainText -Force ))

# Create a resource group
New-AzResourceGroup -Name $resourceGroup -Location $location

# Create a subnet configuration
$subnetConfig = New-AzVirtualNetworkSubnetConfig -Name mySubnet -AddressPrefix 10.0.3.0/24

# Create a virtual network
$vnet = New-AzVirtualNetwork -ResourceGroupName $resourceGroup -Location $location `
  -Name MYvNET -AddressPrefix 10.0.3.0/24 -Subnet $subnetConfig

# Create a public IP address and specify a DNS name
$pipMgr = New-AzPublicIpAddress -ResourceGroupName $resourceGroup -Location $location `
  -Name "mgr$(Get-Random)" -AllocationMethod Static -IdleTimeoutInMinutes 4 -DomainNameLabel "mgr$(Get-Random)"
$pipWorker = New-AzPublicIpAddress -ResourceGroupName $resourceGroup -Location $location `
    -Name "worker$(Get-Random)" -AllocationMethod Static -IdleTimeoutInMinutes 4 -DomainNameLabel "worker$(Get-Random)"

# Create an inbound network security group rule for port 3389 and 80
$nsgRuleRDP = New-AzNetworkSecurityRuleConfig -Name myNetworkSecurityGroupRuleRDP  -Protocol Tcp `
  -Direction Inbound -Priority 1000 -SourceAddressPrefix * -SourcePortRange * -DestinationAddressPrefix * `
  -DestinationPortRange 3389 -Access Allow
$nsgRuleHTTPS = New-AzNetworkSecurityRuleConfig -Name myNetworkSecurityGroupRuleHttps  -Protocol Tcp `
  -Direction Inbound -Priority 1010 -SourceAddressPrefix * -SourcePortRange * -DestinationAddressPrefix * `
  -DestinationPortRange 443 -Access Allow

# Create a network security group
$nsgMgr = New-AzNetworkSecurityGroup -ResourceGroupName $resourceGroup -Location $location `
  -Name myNetworkSecurityGroupMgr -SecurityRules $nsgRuleRDP,$nsgRuleHTTPS
$nsgWorker = New-AzNetworkSecurityGroup -ResourceGroupName $resourceGroup -Location $location `
  -Name myNetworkSecurityGroupWorker -SecurityRules $nsgRuleRDP,$nsgRuleHTTPS

# Create a virtual network card and associate with public IP address and NSG
$nicMgr = New-AzNetworkInterface -Name myNicMgr -ResourceGroupName $resourceGroup -Location $location `
  -SubnetId $vnet.Subnets[0].Id -PublicIpAddressId $pipMgr.Id -NetworkSecurityGroupId $nsgMgr.Id
$nicWorker = New-AzNetworkInterface -Name myNicWorker -ResourceGroupName $resourceGroup -Location $location `
  -SubnetId $vnet.Subnets[0].Id -PublicIpAddressId $pipWorker.Id -NetworkSecurityGroupId $nsgWorker.Id

# Create a virtual machine configuration
$vmConfigMgr = New-AzVMConfig -VMName $mgrName -VMSize Standard_D4d_v4 | `
Set-AzVMOperatingSystem -Windows -ComputerName $mgrName -Credential $cred | `
Set-AzVMSourceImage -PublisherName MicrosoftWindowsServer -Offer WindowsServer -Skus 2019-datacenter-core-with-containers -Version latest | `
Add-AzVMNetworkInterface -Id $nicMgr.Id
$vmConfigWorker = New-AzVMConfig -VMName $workerName -VMSize Standard_D4d_v4 | `
Set-AzVMOperatingSystem -Windows -ComputerName $workerName -Credential $cred | `
Set-AzVMSourceImage -PublisherName MicrosoftWindowsServer -Offer WindowsServer -Skus 2019-datacenter-core-with-containers -Version latest | `
Add-AzVMNetworkInterface -Id $nicWorker.Id

# Create a virtual machine
New-AzVM -ResourceGroupName $resourceGroup -Location $location -VM $vmConfigMgr
New-AzVM -ResourceGroupName $resourceGroup -Location $location -VM $vmConfigWorker
```
4. connect to the mgr vm and run the following commands to validate that the relevant KB is installed, setup the firewall and init the swarm
```
powershell
get-hotfix
New-NetFirewallRule -DisplayName "Allow Swarm TCP" -Direction Inbound -Action Allow -Protocol TCP -LocalPort 2377, 7946 | Out-Null
New-NetFirewallRule -DisplayName "Allow Swarm UDP" -Direction Inbound -Action Allow -Protocol UDP -LocalPort 4789, 7946 | Out-Null
New-NetFirewallRule -DisplayName "Allow HTTPS" -Direction Inbound -Action Allow -Protocol TCP -LocalPort 443 | Out-Null

$content = Invoke-WebRequest -H @{ Metadata = "true" } http://169.254.169.254/metadata/instance?api-version=2017-04-02 -UseBasicParsing
$json = ConvertFrom-Json $content
$ipaddress = $json.network.interface.ipv4.ipAddress.privateIpAddress
Write-Host "Found IP address $ipaddress"
Invoke-Expression "docker swarm init --advertise-addr $ipaddress --default-addr-pool 10.10.0.0/16"
```
5. copy the join command
6. go to the worker vm and run the following commands to validate that the relevant KB is installed, setup the firewall and connect to the swarm, of course your join token will be different
```
powershell
get-hotfix
New-NetFirewallRule -DisplayName "Allow Swarm TCP" -Direction Inbound -Action Allow -Protocol TCP -LocalPort 2377, 7946 | Out-Null
New-NetFirewallRule -DisplayName "Allow Swarm UDP" -Direction Inbound -Action Allow -Protocol UDP -LocalPort 4789, 7946 | Out-Null

docker swarm join --token SWMTKN-1-4y29h72magt5i6g4nwh3099dy6empntfu7ztm3ynljbrbq6q9z-dluq3j2vwvzlsmxw8j4be9jje 10.0.3.4:2377
```
7. go back to the mgr and validate that the worker has joined
```
docker node ls
```
8. create an overlay network
```
docker network create --driver=overlay overlay
```
9. validate that a directly connected IIS container can be reached by running the following command and then accessing port 443 with http (!) on the mgr VM from your laptop --> this should give you the standard IIS page with the blue rectangles and proves that no firewall or anything else is blocking access to port 443
```
docker run -p 443:80 -d --rm --name iis mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
```
10. remove the directly connected container
```
docker rm -f iis
```
11. run the following to work around a traefik bug, create an overlay network and create a file docker-compose.yml which describes a swarm setup with traefik, portainer and portainer agents. Replace $externaldns with the external DNS name of the mgr and $email with your email address
```
New-Item -Path c:\le -ItemType Directory | Out-Null	
New-Item -Path c:\le\acme.json | Out-Null
docker network create --driver=overlay traefik-public

@"
version: '3.7'

services:
  traefik:
    image: traefik:2.2-windowsservercore-1809
    command:
#      - --log.level=DEBUG
      - --api.dashboard=true
      - --providers.docker.swarmMode=true
      - --providers.docker.network=traefik-public
      - --providers.docker.exposedbydefault=false
      - --providers.docker.endpoint=npipe:////./pipe/docker_engine
      - --entrypoints.websecure.address=:443
      - --certificatesresolvers.myresolver.acme.tlschallenge=true
      - --certificatesresolvers.myresolver.acme.email=$email
      - --certificatesresolvers.myresolver.acme.storage=c:/le/acme.json
      - --serversTransport.insecureSkipVerify=true
    volumes:
      - source: 'C:/le'
        target: 'C:/le'
        type: bind
      - source: '\\.\pipe\docker_engine'
        target: '\\.\pipe\docker_engine'
        type: npipe
    ports:
      - 443:443
    deploy:
      placement:
        constraints:
          - node.role == manager
      labels:
        - traefik.enable=false
        - traefik.http.routers.api.entrypoints=websecure
        - traefik.http.routers.api.tls.certresolver=myresolver
        - traefik.http.routers.api.rule=Host(``$externaldns``) && (PathPrefix(``/api``) || PathPrefix(``/dashboard``))
        - traefik.http.routers.api.service=api@internal
        - traefik.http.services.api.loadBalancer.server.port=8080
    networks:
      - traefik-public

  agent:
    image: portainer/agent:windows1809-amd64
    environment:
      AGENT_CLUSTER_ADDR: tasks.agent
    volumes:
      - source: '\\.\pipe\docker_engine'
        target: '\\.\pipe\docker_engine'
        type: npipe
      - source: 'C:/ProgramData/docker/volumes'
        target: 'C:/ProgramData/docker/volumes'
        type: bind
    networks:
      - agent-network
    deploy:
      mode: global
      placement:
        constraints:
          - node.platform.os == windows

  portainer:
    image: portainer/portainer:windows1809-amd64
    command: -H tcp://tasks.agent:9001 --tlsskipverify
    volumes:
      - portainer-data:c:/data
    networks:
      - agent-network
      - traefik-public
    deploy:
      placement:
        constraints:
          - node.role == manager
      labels:
        - traefik.enable=true
        - traefik.http.routers.portainer.rule=Host(``$externaldns``) && PathPrefix(``/portainer/``)
        - traefik.http.routers.portainer.entrypoints=websecure
        - traefik.http.routers.portainer.tls.certresolver=myresolver
        - traefik.http.routers.portainer.service=portainer@docker
        - traefik.http.services.portainer.loadBalancer.server.scheme=http
        - traefik.http.services.portainer.loadBalancer.server.port=9000
        - traefik.http.middlewares.portainer.stripprefix.prefixes=/portainer
        - traefik.http.routers.portainer.middlewares=portainer@docker
    depends_on:
      - traefik

networks:
  agent-network:
    attachable: true
  traefik-public:
    external: true

volumes:
  portainer-data:
"@ | Out-File docker-compose.yml
```
12. deploy that as stack
```
docker stack deploy -c .\docker-compose.yml mystack
```
13. validate that you can reach portainer using https://< external dns name >/portainer/
14. reboot the machine
```
restart-computer -force
```
15. try to reach portainer again, which for me fails, so the reboot somehow breaks this
16. remove the stack and try the directly connected container again, which should still work
```
docker stack rm mystack
docker run -p 443:80 -d --rm --name iis mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
```
16. remove the directly connected container and try the stack deployment again
```
docker rm -f iis
docker stack deploy -c .\docker-compose.yml mystack
```
18. try to reach portainer, which for me now works again, so the directly connected container somehow seems to fix it
19. reboot once more to break it and then try to remove the stack and deploy it again
```
docker stack rm mystack
docker stack deploy -c .\docker-compose.yml mystack
```

This for me still fails, so removing and adding doesn't seem to work

In order to validate that this is caused by the fix (or at least one of them), do the exact same, but instead of using `-Version latest` in lines 49 and 53 of the first script (step 3), now use `-Version "17763.1158.2004131759"`. In this case, step 15 doesn't break

Let me know if you need anything else or the steps don't work for you

**Comment by veepee78** [comment]
> @tfenster yes please. Unfortunately this is not reproducing for me and others that have tried on a Windows Server 2019 machine with latest hotfix. Perhaps can someone share a sample .yaml file as well in case there is something different in the configuration?

Do you mean by hotfix the 2020-06 cumulative update?

**Comment by djarvis** [comment]
@tfenster This is the scenario that I have been citing.  It reproduces the problem with KB4551853 installed:

On Windows Server 2019 (any flavor) with KB4551853 installed, install docker:

* PS> Install-Module -Name DockerMsftProvider -Repository PSGallery -Force 
* PS> Install-Package -Name docker -ProviderName DockerMsftProvider -Force 
* PS> Restart-Computer 
* PS> Invoke-WebRequest "https://github.com/docker/compose/releases/download/1.25.3/docker-compose-Windows-x86_64.exe" -UseBasicParsing -OutFile $Env:ProgramFiles\Docker\docker-compose.exe 

Once docker and docker-compose is installed we can run the sample web container without docker swarm to verity we are getting an output:
  * docker run -d -p 80:80 djarvis8/simpleweb-win
  * From a web browser or curl, pull down the simple web page with which the container responds.  Do this from a remote computer, assuring port 80 is allowed through firewalls.  You should see "You have reached this simple web page".  All is well.

Now we will launch the container within a swarm and watch it fail:

* docker stop <<container id>>
* docker swarm init --advertise-addr={{IP ADDRESS}}
* docker stack deploy -c docker-compose.yaml testapp

docker-compose.yaml:

```
version: '3.7'
services:      
  simpleweb:
    image: "djarvis8/simpleweb-win:latest"
    networks:
      - testnet
    ports:
      - target: 80
        published: 80
        protocol: tcp
        mode: ingress

networks:
  testnet:
    driver: overlay
    attachable: true

```

* Again from a remote web browser or curl, try to pull down the web page.  Expected result: the web page.  Actual result: nothing.

**Comment by djarvis** [comment]
@daschott I posted a failing scenario here but deleted it since I need to revise it a bit.  I'll have a failing scenario shortly...

**Comment by djarvis** [comment]
@tfenster @daschott This is the scenario that I just tested.  It reproduces the problem with KB4551853 installed.  It does seem to have to do with container-to-container communication in a swarm network:

On Windows Server 2019 (any flavor) with KB4551853 installed, install docker:

* PS> Install-Module -Name DockerMsftProvider -Repository PSGallery -Force 
* PS> Install-Package -Name docker -ProviderName DockerMsftProvider -Force 
* PS> Restart-Computer 
* PS> Invoke-WebRequest "https://github.com/docker/compose/releases/download/1.25.3/docker-compose-Windows-x86_64.exe" -UseBasicParsing -OutFile $Env:ProgramFiles\Docker\docker-compose.exe 

What we will do is run two simple containers.  Simpleweb when hit will return a simple HTML page.  Simpleweb2 when hit will internally fetch http://simpleweb and report back the simple web page. We will do this first outside of swarm to verify this all works:

  * docker-compose up

docker-compose.yaml:

```
version: '3.7'
services:      

  simpleweb:
    image: "djarvis8/simpleweb-win:latest"

  simpleweb2:
    image: "djarvis8/simpleweb-win:latest"
    ports:
      - target: 80
        published: 80
        protocol: tcp
        mode: ingress
    environment:
      Url: http://simpleweb



```
  * From a web browser or curl, pull down the simple web page with which the container responds.  Do this from a remote computer, assuring port 80 is allowed through firewalls.  You should see success:
```

Fetching http://simpleweb
Result: OK
Content: <HTML><BODY><B>You have reached this simple web page</B></BODY></HTML>
```


Now we will launch the container within a swarm and watch it fail:

* docker swarm init --advertise-addr={{IP ADDRESS}}
* docker stack deploy -c simple.yaml testapp

simple.yaml:

```
version: '3.8'
services:      

  simpleweb:
    image: "djarvis8/simpleweb-win:latest"
    networks:
      - testnet


  simpleweb2:
    image: "djarvis8/simpleweb-win:latest"
    networks:
      - testnet
    ports:
      - target: 80
        published: 80
        protocol: tcp
        mode: ingress
    environment:
      Url: http://simpleweb


networks:
  testnet:
    driver: overlay
    attachable: true
```

* Again from a remote web browser or curl, try to pull down the web page.  You should see failure:

```
Fetching http://simpleweb
Exception: One or more errors occurred. (A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.)
```

**Comment by xargon180** [comment]
Is there a workaround without removing the update? Our internal company security policy prevents me from uninstalling of KB4551853.

**Comment by tfenster** [comment]
@daschott Any feedback on this? Can you repro the issue now?

**Comment by daschott** [comment]
@tfenster To give an update, product group is still actively working together with support teams to analyze repro traces collected from a environment.

@xargon180 What is the scenario you are targeting and looking for a workaround? Depending on the scenario, you *may* be able to use hostPort publishing mode. But I would need to hear more to confirm.

**Comment by xargon180** [comment]
@daschott Thanks for the tip. Unfortunately my services could still not communicate with each other in host mode. Temporarily I rolled out my stack with docker-compose to a single server. I hope that there will be a real solution soon :pray:.

**Comment by djarvis** [comment]
@xargon180 Same situation here, we're going out with docker-compose as we see no other option right now.  

@daschott was the repo description above of any value?

**Comment by jabteles** [comment]
I'm afraid this is also happening with [KB4561608](https://support.microsoft.com/pt-pt/help/4561608/windows-10-update-kb4561608).

I had the same issue again, services not reachable from other services.

Noticed this KB was installed. Uninstalled, started working. They are both Cumulative Updates, from May and June.

**Comment by mpnewcomb** [comment]
> I'm afraid this is also happening with [KB4561608](https://support.microsoft.com/pt-pt/help/4561608/windows-10-update-kb4561608).
> 
> I had the same issue again, services not reachable from other services.
> 
> Noticed this KB was installed. Uninstalled, started working. They are both Cumulative Updates, from May and June.

Confirmed, nothing could get in. Removed this cumulative update and boom, started working. I will restore to a fresh Windows install adding cumulative updates starting from April going backwards until I find which cumulative update broke it.

**Comment by mpnewcomb** [comment]
April 21, 2020 [KB4550969](https://support.microsoft.com/en-us/help/4550969/windows-10-update-kb4550969) (OS Build 17763.1192), broke swarm ingress. I uninstalled this update and it went back to working.
April 14, 2020 [KB4549949](https://support.microsoft.com/en-us/help/4549949/windows-10-update-kb4549949) (OS Build 17763.1158), confirmed working.

**Comment by daschott** [comment]
Thanks everyone for the repro descriptions. Team is still investigating this issue. Please expect an update early next week.

**Comment by daschott** [comment]
There was a regression found in the port pool optimizations that shipped with the cumulative updates.

Unfortunately, the fix for this regression is not available yet on any public Windows releases. Microsoft is working on generating a new patch for Windows Server 2019. ETA for creating this patch is ~1 week. 

@djarvis Please reach out to Microsoft customer support to receive this patch as soon as it's ready for sharing.

**Comment by simmohall** [comment]
awesome! looking forward to being able to get our servers all updated with latest patches again :)

**Comment by djarvis** [comment]
@daschott Awesome, sounds good!

**Comment by daschott** [comment]
The patch has been generated and is ready for sharing with impacted users that request it. Please reach out to your Microsoft customer service contact to ask for a patch containing "Separate SNAT pools for each source VIP".

Once there is enough confidence and confirmation from users that the patch indeed resolves this regression, I will share which cumulative update would include this and when it will be released.

**Comment by xargon180** [comment]
Any news? Has anyone tested the secret patch yet?

**Comment by djarvis** [comment]
@xargon180 We're currently testing it out.

**Comment by ncresswell** [comment]
can someone please advise when we should expect this to be public domain? It basically renders production servers and services offline.. i ended up rebuilding my Windows Hosts with an old ISO and blocking windows updates to work around this.

**Comment by simmohall** [comment]
yeah, this is becoming a bigger problem given that our servers are unable to be patched since April! @daschott do we have a patch number or reference? Our operations team haven't been able to get a hold of this patch from our support partners yet.

**Comment by djarvis** [comment]
@daschott The patch works for us- our containers are able to communicate again.

**Comment by immon4ik** [comment]
@djarvis hi, please tell me how to get this patch?

**Comment by immon4ik** [comment]
For me, installing the cumulative update KB4558998, restarting the nodes and rotating the roles solved the problem.

**Comment by xargon180** [comment]
@immon4ik For me, patch KB4558998 makes no differences. Swarm still does not work.
@daschott Is there an official statement in which patch the bug fix is included?

**Comment by daschott** [comment]
@xargon180 No publicly available patch has the fix included yet. Only private patch that can be requested through Microsoft Support.

The tentative ETA for the cumulative update (public patch) is 15th September 2020.

**Comment by veepee78** [comment]
Any idea has this patch been released in september cumulative? KB4570333
https://support.microsoft.com/en-us/help/4570333/windows-10-update-kb4570333 that does not say anything about docker.

**Comment by simmohall** [comment]
I was hopeful too, but i just spun up 2 vms in azure, applied september patches and still can't connect multiple services together :( pretty disappointing this is taking so long given that this is a pretty fundamental regression break to docker swarm on a Win2019 environment. Something that the MS docs say is still supported.

**Comment by daschott** [comment]
@veepee78 @simmohall Sorry to hear you are disappointed but KB4570333 does not include the patch. The tentative ETA I shared in a [comment above](https://github.com/moby/moby/issues/40998#issuecomment-670090190) is referring to the next cumulative update (generally released on the 3rd Tuesday each month). If you need the fix before that, please reach out to Microsoft support and they will be able to share it with you before the public release.

**Comment by lindback** [comment]
> @immon4ik For me, patch KB4558998 makes no differences. Swarm still does not work.

@immon4ik For us, the patch makes exposed ports in swarm accessible again, but we still have an issue when we have traefik as a LB in our swarm, Microsoft is looking into this issue as well.

**Comment by pralobo** [comment]
@daschott Does the patch KB4577069 (https://support.microsoft.com/en-us/help/4577069) fix the issue? If not, when is the ETA?

**Comment by simmohall** [comment]
I tried downloading the KB4577069 and it's fixed my issue. Containers within the swarm overlay can again communicate with each other :) :) :)

**Comment by olljanat** [comment]
@daschott before we close this one. Can you confirm that regression test for this case have now included to part of Windows build/test process? Those of us who have been running Windows containers since srv 2016 RTM release have seen overlay networking breaking too many times because of broken Win updates.

I'm working on to enable more integration tests for Windows on Moby side which hopefully will avoid overlay breaking because changes made on here but as we do not have visibility to what happens for Microsoft side we can just wish better code quality...

**Comment by veepee78** [comment]
Installed kb...069 and still cannot connect to overlay attached container by opened port, nor from container to external world. Tested with IIS container. Also ping does not work to default gateway inside container. As control, on same host when container is connected to default nat network, everything works.

**Comment by olljanat** [comment]
FYI. #41434 contains interesting finding that network name(s) which comes before "ingress" in alphabetic order are especially problematic.

When I looks `docker inspect <container id>` inspection from those cases I can see that network **hotel** is really listed before ingress:
```yaml
"Networks": {
	"hotel": {
		"IPAMConfig": {
			"IPv4Address": "10.0.1.9"
		},
		"Links": null,
		"Aliases": [
			"32da675f7e61"
		],
		"NetworkID": "z66q3jnwc4erladpfhxvvz8mh",
		"EndpointID": "6ba25843a960e2ae140b807cd7ab33453a837e75f02d9988699125613998771b",
		"Gateway": "",
		"IPAddress": "10.0.1.9",
		"IPPrefixLen": 24,
		"IPv6Gateway": "",
		"GlobalIPv6Address": "",
		"GlobalIPv6PrefixLen": 0,
		"MacAddress": "00:15:5d:8b:e7:63",
		"DriverOpts": null
	},
	"ingress": {
		"IPAMConfig": {
			"IPv4Address": "10.0.0.12"
		},
		"Links": null,
		"Aliases": [
			"32da675f7e61"
		],
		"NetworkID": "qhsdoekcz77e1mzdkhdm4ogrr",
		"EndpointID": "745ea40e28d19982ab888c1c39023910bc32d2823f470747ad8253ca8e2edefa",
		"Gateway": "",
		"IPAddress": "10.0.0.12",
		"IPPrefixLen": 24,
		"IPv6Gateway": "",
		"GlobalIPv6Address": "",
		"GlobalIPv6PrefixLen": 0,
		"MacAddress": "00:15:5d:1d:00:2f",
		"DriverOpts": null
	}
}
```
so probably networks gets created to HNS also same order?

I also noticed that if on Linux containers ingress network is not listed on spec at all. Not sure if that is just because of different implementation or is there actually some bug on Docker side...

**Comment by thaJeztah** [comment]
^^ @arkodg PTAL

**Comment by olljanat** [comment]
Some more findings from my debugging and also some workarounds which I would like to see others to confirm. All these tests was done with Win 2019 with KB4577069 installed.

### Scenario: Single node swarm, Windows Server 2019 as only manager
When swarm is initialized "ingress" network gets created to Docker but only partly to HNS side.
Configuration will be finalized when you create first service. So based on scenario on #41434 you must create one service is uses network "juliett" first. After that you can add another one which uses network "hotel" and both of those will works just fine.

### Scenario: Production grade cluster, Linux servers are managers, Windows Server 2019 as worker
When you join Windows to existing swarm as worker "ingress" gets fully configured to HNS side too and you can then use networks with any name you like.

### General findings
On both scenarios, if you reboot Windows node "ingress" will stop working.

You can avoid that by stopping Docker service first and only then reboot server. 

If your server is already on this state you can either:
- Leave from swarm and join back to it.
- Remove problematic network from HNS by
  - Stop Docker
  - Check with command `hnsdiag list networks` which networks except "nat" there is.
  - Remove those networks using command `hnsdiag delete networks <network GUID>`
  - Start Docker again.

## Proposal
Based on this I propose that we (will look about it but I'm not sure if I find how to do it) update Docker logic on way that:
- Docker will fully configure "ingress" even on single node swarm after swarm init.
- Stop unconfiguring "ingress" when stopping daemon.
- If there is any other "overlay" networks than "ingress" when daemon starts we should drop those (they will get re-created when swarm schedules new containers to start).

**Comment by simmohall** [comment]
@olljanat  interesting findings. After my initial excitement, i've discovered similar issues. The initial swarm works with the KB4577069. However after some stack deployments after a day or two, they've lost the connections again :(

My setup is 1 linux manager, 2 windows manager, and another windows worker. My network names alphabetically come after ingress. 'proxy', 'uat', 'sit'.

**Comment by alexftr** [comment]
For us Swarm stacks work with the KB4577069 only until restart of the server. After restart containers are not able to reach each other. 
It only works again if I remove all stacks then restart the server and then create all stacks again.

**Comment by veepee78** [comment]
I can't repeat that. Initiated swarm, created overlay net "john" and attached iis container to it. Can't connect anything from within the container nor to published port.

Are you saying that you have issues only with container to container connections? I can't connect anything from overlay network.
I have this behavior on two different machines.

1809 server with latest patches, 190311 docker

**Comment by olljanat** [comment]
@simmohall @alexftr @veepee78 OK. Good to know. There looks to be lot of variables on this puzzle so very hard to make more stable. Especially if Microsoft does not provide more info that what actually have been changed on HNS and how it was tested.

However I did bigger scenario testing now with more production like configuration where Traefik top of Linux is used to publish services. You can find those stacks from https://github.com/olljanat/docker-issue-tracking That one I was not able to break with couple of reboots what I did for those Windows nodes. If you find bugs or you have improvement requests/ideas/etc to that one you can also open issues to there.

With that said, let me clarify that on this discussion we are interested only about issues which created KB4551853 and still exists after KB4577069. Anything outside of those should be handled on other issues.

Currently known are:
* That strange behavior with network names #41434 on single Windows node cluster
* Connectivity issue from Windows to Linux #41354 unless `endpoint_mode=dnsrr` setting is used.
* Fact that networks breaks on reboots more often than earlier.

Afaik, any of those wasn't problem before KB4551853. Also note that I have been trying to setup swarm to Moby CI but that work is currently on hold because of the fact that HNS configuring/reconfiguring breaks CI and make it very unstable https://github.com/moby/moby/pull/38432#issuecomment-696009842

**Comment by turric4n** [comment]
Same problem with one 3 physical Windows 2019 with docker 19.03.12 Swarm mode.

With KB4577069 installed.

**Comment by s13n4** [comment]
KB4577069 breaks ingress working on 17763.1397

**Comment by xargon180** [comment]
Unfortunately the current patch KB4577069 still does not solve the problem.

**Comment by simmohall** [comment]
@daschott  is their any news on this issue? Can we hope for a fix in October patches!

**Comment by ncresswell** [comment]
I ended up rebuilding my Windows hosts and disabling Windows Update so as to NOT get any “breaking” updates… this whole shambles just shows how fragile Docker on Windows Server is, and its quite a risk to businesses using Windows containers in Production.

**Comment by mkoziel2000** [comment]
Just so this issue includes Kubernetes folks.  I've had the same problems with DNS connectivity with PODS when KB4570333 was applied.  Backing it back out allowed the kubernetes windows worker nodes to start working again when it came to DNS resolution.

The trials and tribulations of getting windows nodes working with containers has been nothing short of frustrating.  I hope Microsoft is investing more into this as containerization is a must have like VMs where back in the 2000's...but people are going to abandon WIndows OS's if these breaking changes keep happening.

**Comment by redjockey** [comment]
Installed kb4577063 on 2 freshly installed windows server core 2004 VMs. The 2004 RTM was already broken, so could finally start testing. Creating the swarm works again and the traefik service is reachable again on the ports that are defined. However if i update/redeploy the traefik service it stops working. Only when i restart the docker host it is reachable again on those ports.

**Comment by xargon180** [comment]
@djarvis Did the inofficial "Separate SNAT pools for each source VIP" patch still work four you after installation of patch KB4577069 ?
Unfortunately Microsoft enterprise support told me that they do not have a patch for "Separate SNAT pools for each source VIP".

**Comment by simmohall** [comment]
sooo....anyone tried the October 13 patch KB4577668? This has become a joke - why can't this issue be rolled back or fixed in 6 months! Next priority is migrating all apps to .net core so they can run on linux.

**Comment by daschott** [comment]
Please await KB4580390. It is estimated to come out on 10/20/20 at 10:00 am and contains 2 important fixes for regressions being mentioned in this discussion to containers on Windows Server 2019.

@redjockey Can you share details about the issues you are seeing on Windows Server, version 2004? Containers in Windows Server, version 2004 is used successfully in production at large scale by users.

@mkoziel2000 Indeed, setting up Kubernetes on Windows Server yourself is challenging. We acknowledge this and recommend [AKS HCI](https://docs.microsoft.com/en-us/azure-stack/aks-hci/) which enables creation of Kubernetes clusters with best practices conveniently using Windows Admin Center UI or command line. Many of the workarounds and additional work on Kubernetes is due to lack of privileged containers support on Windows. This is coming with 1.21 Kubernetes release and will simplify the process if you want to spin up Kubernetes yourself. In the meantime, if you want to spin up Kubernetes yourself there is also a blog post [here](https://techcommunity.microsoft.com/t5/networking-blog/troubleshooting-kubernetes-networking-on-windows-part-1/ba-p/508648) describing how to troubleshoot networking issues on Windows Server.

@simmohall Sorry to hear your pain and frustrations with the cumulative updates process. As you know, containers is a rapidly moving space. Despite our best efforts including months of internal testing and validation efforts by impacted customers, it is clear we still had gaps in our testing when shipping back a requested feature from SAC to LTSC releases. In this case, there was a fix partially missing that existed for a scalability enhancement feature that originally shipped with Windows Server, version 2004. Of course anytime there are regressions relevant teams adjust testing procedures accordingly.

Our recommendation continues to be to use the latest Windows Server version for containers features such that you get the most mature implementation.

**Comment by daschott** [comment]
@xargon180 can you tell your support contact to reach out to me? There are a few users that requested the private patch and confirmed it had resolved their issue. If you're comfortable you can also reach me over Slack @daschott on https://kubernetes.slack.com/messages/sig-windows and I can share the patch with you directly.

**Comment by simmohall** [comment]
Thanks for the update @daschott much appreciated! I guess the main frustration has not been knowing if and when it would be fixed. Having a more concrete idea of when a fix is expected will help to mitigate operations security concerns of unpatched servers.
Are you advising the recommended approach is to move to the SAC rather than LTSC releases for running containers in the windows environment?

**Comment by redjockey** [comment]
@daschott What do you want to know specifically? Before the preview patch, i could run docker on 2004, but not in swarm mode. When i did a docker swarm init things stopped working. The server would loose its static ip and become unreachable. With a DHCP server added, it would switch from static to DHCP. But any services on it with open ports werent reachable at all.

The preview patch kb4577063, fixed most of those issues. Can setup the swarm again, the static IP stays as is during the swarm init. I have traefik setup on port 80, 443, and 8080. There are about 15 other containers serving various websites behind it.  The sites and traefik are reachable again. Redeploying/updating the website containers doesnt cause any issues. Any changes get picked up by traefik just fine.

I forgot some labels we use for logging and when i added those the traefik compose file and redeployed. Things stopped working. The ingress ports for traefik 80,443,8080 were gone, i did a port scan from my own pc on the docker host to verify. Only after a reboot of the docker host, it started to work again with the updated config. If i do a docker stop on the traefik container, it gets restarted again by the service. But after that its unreachable aswell. Until the host gets rebooted. There seems to be a problem with reassigning the ingress ports that are being used.

**Comment by veepee78** [comment]
Installed all updates and KB4580390, just a small improvement, IIS container in ingress network, I'm able to connect to iis through opened port. However cannot still connect to anything from inside the container. 

I think I'm quite done with playing with windows docker..

**Comment by daschott** [comment]
[KB4580390](https://support.microsoft.com/en-us/help/4580390) was released containing fixes for the prior regression and @djarvis issue. Can you try it out and confirm whether the original issue still reproduces or not?

@redjockey can you file a separate issue for losing traefik ingress ports until reboot with the details you described?
@veepee78 Can you describe what cmdlet/program did you use to establish connectivity and what is the destination? Are you unable to `Invoke-webrequest google.com` or what are you trying to connect to? Keep in mind that overlay networks on Windows do not allow connecting to the host network, although external Internet connectivity is enabled.

**Comment by veepee78** [comment]
@daschott I created single host swarm, run iis image as service and I used ping inside container. Pinged few internet addresses, including google.com and also our internal network addresses. No response.  This used to work early this year. 

If I run the container in default nat network, everything responds.

**Comment by daschott** [comment]
@veepee78 Instead of ping, can you try to use `curl` or `invoke-webrequest` for google.com? Most applications use TCP or UDP protocol, but ping uses ICMP. Unfortunately ICMP is not going to work from inside a container on Windows Server 2019.
https://docs.microsoft.com/en-us/virtualization/windowscontainers/container-networking/network-drivers-topologies#overlay-network-driver

**Comment by simmohall** [comment]
Thanks @daschott ! 
I've installed the patch KB4580390 in our test cluster and have been watching over the past few days. So far it's looking much better. 
Comms between containers is back working, and even after multiple service updates the connections are still remaining. 
I'll give it a another week this time before I jump the gun and push to our PROD clusters :)

**Comment by veepee78** [comment]
@daschott Thanks, testing with tcp seems to work. Has that icmp rule changed as I recall testing connection from container with ping earlier with success?

So, this seems actually work now. Though only one service at the moment tested and confirmed working.

**Comment by tfenster** [comment]
@daschott or anyone else: Would you have any idea when that patch appears in the standard Azure images for VMs? I have a fully automated setup which I would like to test, but the images don't have the KB yet
![image](https://user-images.githubusercontent.com/15346084/97007779-379d1b80-1542-11eb-87be-8d5624a7c474.png)

**Comment by olljanat** [comment]
@tfenster [KB4580390](https://support.microsoft.com/en-us/help/4580390/windows-10-update-kb4580390) is currently marked as "preview" on KB article so I assume that MS will create new VM images to Azuze next time after November patch Tuesday before that you need install those using methods described on KB.

**But** I must say that running Windows Swarm nodes on Azure is IMO totally nonsense. If you are already using Azure then it much simpler just use [AKS with Windows](https://docs.microsoft.com/en-us/azure/aks/windows-container-cli) (and example Kompose or Portainer 2.0 to simplify migration). That one is actively developed and don't suffer lag of support from Microsoft side like swarm support does.

Windows Swarm nodes only makes sense for those who are already using some other platform that Azure and who do not want struggle which comes with maintaining Kubernetes.


@daschott Looks that KB4580390 helps but I cannot see anything HNS related either on KB4551853 or KB4580390 release notes so can you please shortly describe that what actually caused this issue and what is done to make sure that it does not came back on future upgrades?

**Comment by Zordaxy** [comment]
Still doesn't work for me.
Last installed patch is KB4580390.

Steps:
1. docker swarm init --advertise-addr {VM's public IP address}
2. docker service create --name iis --publish 8080:80 mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
3. curl http://{VM's IP address}:8080 
results in "curl : Unable to connect to the remote server".

So I still can't access the service from the outside.

Some notes:
1. "docker exec {container ID} curl http://localhost:80" - gives vaid result. It means iis works fine inside container.
2. If I try to launch something else in port 8080 - get notification that port is used. Once I stop service - port is available again.
3. In "mode=host" service works well. For example "docker service create --name ziis --publish mode=host,target=80,published=4030 mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019" works fine in assigned port
4. "netstat -ano" - does not show anythind on assigned port

**Comment by ncresswell** [comment]
I actually had the inverse after applying this patch.

If i create a service on the overlay network and publish ports as “mode=host” the port is opened, but it doesnt work.

If i create a service on the overlay network and publish ports via ingress, all works fine.

Weird.

**Comment by simmohall** [comment]
I'm happy to consider this issue as closed - the original issue that was opened from 2020-05 KB4551853 update, causing containers not being able to communicate on the swarm ingress network - now has been fixed with the KB4580390 patch. I've had no issues on our swarm for over a week. Thanks all 👍

**Comment by sam9291** [comment]
I've been struggling on this problem on and off for the past month, I can confirm KB4580390 fixes the issue on my side on a cluster with 2 nodes (windows server 2019), when testing with the previous patch KB4577668 on an overlay network across the two nodes, they where not working. With both nodes updated with KB4580390 it is! Thanks for the help!

![image](https://user-images.githubusercontent.com/5232883/97945799-9b9bbd00-1d56-11eb-895e-f744c83e4b2b.png)

**Comment by turric4n** [comment]
Confirmed, KB4580390 solved the problem with docker swarm overlay network.

**Comment by terrpan** [comment]
After installing KB4580390 we're experincing a lot of on and off errors with the HNS. Everything can run without any issues for a couple of hours and all of sudden everything crashes and we have clean out HNS networks and reboot the server.

Anyone else experiencing the same?

**Comment by sam9291** [comment]
We've been running the patch on 3 servers (1 cluster of 1 node and 1 cluster of 2 nodes) and they seems to have been pretty stable in the past 2 weeks on my end.

**Comment by mkoziel2000** [comment]
We've been running with one node in the kubernetes cluster patched with the Nocv cumulative...no issues with network connectivity.  seems the issue introduced in september appears to have been fixed with the oct 20 patch.

**Comment by thaJeztah** [comment]
Thanks all! Let me close this issue as it looks like it's resolved

**Comment by daschott** [comment]
@terrpan What errors are you seeing? Can you file a separate issue for this? Can you try to remove HNS policylist before deleting networks? It may help avoid the reboot.

@veepee78 Yes, this indeed has changed in Windows Server 2019 and recent Docker versions. I believe Windows Server vNext next year will not have this issue and `ping` should be able to work. Regardless, most applications typically use TCP/UDP and that should work now.
